Iotssl 1293 add ca suppress feature

yanesca opened the pull request:
According to RFC5246 the server can indicate the known Certificate
    Authorities or can constrain the aurhorisation space by sending a
    certificate list. This part of the message is optional and if omitted,
    the client may send any certificate in the response.
    
    The previous behaviour of mbed TLS was to always send the name of all the
    CAs that are configured as root CAs. In certain cases this might cause
    usability and privacy issues for example:
    - If the list of the CA names is longer than the peers input buffer then
      the handshake will fail
    - If the configured CAs belong to third parties, this message gives away
      information on the relations to these third parties
    
    Therefore we introduce an option to suppress the CA list in the
    Certificate Request message.
    
    Providing this feature as a runtime option comes with a little cost in
    code size and advantages in maintenance and flexibility.

mpg commented on library/ssl_srv.c:
```diff
@@ -2588,36 +2588,45 @@ static int ssl_write_certificate_request( mbedtls_ssl_context *ssl )
     total_dn_size = 0;
-    while( crt != NULL && crt->version != 0 )
+
+#if defined(MBEDTLS_SSL_SRV_C)
```
Since we're in `ssl_srv.c` I think this is redundant (same for the one on the closing brace obviously).

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -185,6 +185,9 @@
 #define MBEDTLS_SSL_PRESET_DEFAULT              0
 #define MBEDTLS_SSL_PRESET_SUITEB               2
 
+#define MDEDTLS_SSL_SUPPRESS_CA_LIST_ENABLED     1
```
Please avoid double negatives. Make the constants e.g. `MBEDTLS_SSL_SEND_CA_LIST` and `MBEDTLS_SSL_OMIT_CA_LIST`. Or a boolean “send CA list?” if defaulting to a nonzero representation is acceptable for binary compatibility.

mpg replied:
Historical note: in the past we tended to favor a zero representation of the default as `memset(0)` was sort-of the official way to initialize structures. Now that `mbedtls_foo_init()` is generalized, I don't think there is a reason anymore to favor zero representation for the default value.

gilles-peskine-arm commented on programs/ssl/ssl_server2.c:
```diff
@@ -124,6 +124,7 @@ int main( void )
 #define DFL_MAX_VERSION         -1
 #define DFL_ARC4                -1
 #define DFL_AUTH_MODE           -1
+#define DFL_SUPPRESS_CA         MDEDTLS_SSL_SUPPRESS_CA_LIST_DISABLED
```
Don't be unafraid of double negatives, again.

yanesca replied:
Oh, I see what you did there :)

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -1893,6 +1893,22 @@ run_test    "Authentication: client no cert, ssl3" \
+             key_file=data_files/server6.key" \
+            0 \
+            -S "requested DN"
+
```
These two test cases test the server behavior with a cooperating client, we should also test the client behavior (verify that it doesn't send a non-advertised certificate, and that it will send a certificate if the list from the server is empty), and the server behavor with a non-cooperating client (client that sends a certificate with a non-advertised CA).

mpg replied:
IIRC currently a client can only have 0 or 1 certificate provisioned, not a longer list, so our client implementation currently just ignores the CA list and sends its certificate iff it has one. So, while I agree about adding a test that the client sends a cert when the server CA list is empty, I think the other test will show that our client indeed can send a cert for non-advertised CA.

The RFC doesn't seem to explicitly state what a client is supposed to do when it only has a certificate that doesn't match the non-empty list. But I remember discussions on the TLS mailing list about similar issues (no certificate matching signature_algorithms) and it seemed to me there was a consensus that sending the cert anyway was OK: the peer has to be prepared for than anyway, and is then free to react however it likes (accept the cert anyway (unlikely), continue the handshake considering the peer as _not_ authenticated, abort the handshake).

While the last test (client sends a non-trusted cert) is not about new behavior introduced by this PR, I'm all in favor of taking that occasion to include it if we didn't test that case already (I think/hope we do, but let's make sure).

yanesca commented:
Thank you both for the fast review and the valuable feedback!

I renamed the variables and the macros to use the expression `CERT_REQ` instead of `SUPPRESS`. This way we can avoid the double negative.

Also I have added two tests:
- I have checked and the tests we already have are only testing the case when the client certificate is invalid and not the case when the issuer is not trusted, so I added a test for this case
- Also I added a test for the case when the CA list is suppressed and the client sends an untrusted certificate

gilles-peskine-arm commented on ChangeLog:
```diff
@@ -1,5 +1,12 @@
+Feature
+   * Add a new configuration option to 'mbedtls_ssl_config' to enable
+     suppressing the CA list in Certificate Request messages. The default
+     behaviour has not changed. Namely every configured CAs name is included.
```
nitpick: “… has not changed<strong>, n</strong>amely every …”

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2031,6 +2038,19 @@ void mbedtls_ssl_conf_extended_master_secret( mbedtls_ssl_config *conf, char ems
 
+#if defined(MBEDTLS_SSL_SRV_C)
+/**
+ * \brief          Disable or enable CA list suppression
```
Thanks for changing the name of the constants, please update the documentation to match (e.g. here “Whether to send a list of acceptable CAs in CertificateRequest messages (default: do send)”).

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -1893,6 +1910,34 @@ run_test    "Authentication: client no cert, ssl3" \
 
+# Tests for CA list suppression
+
+run_test    "Authentication: CA suppression disabled (default)" \
```
Please update the test descriptions (e.g. “Authentication: send CA list in CertificateRequest (default)”, “Authentication: suppress CA list in CertificateRequest” or “Authentication: do not send CA lis in CertificateRequest”).

sbutcher-arm commented:
Reviewed, approved, tested and merged.
